Add BedJet BLE climate component

jhansche opened the pull request:
# What does this implement/fix? 

Support for BedJet (V3) climate device over BLE.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** n/a

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1508

## Test Environment

- [X] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
ble_client:
  - mac_address: 11:22:33:aa:bb:cc
    id: my_bedjet

climate:
  - platform: bedjet
    name: "My BedJet Fan"
    ble_client_id: my_bedjet
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

jhansche committed:
Add BedJet BLE climate component

jhansche committed:
Resolve github checks

jhansche commented on esphome/components/bedjet/bedjet.h:
```diff
@@ -0,0 +1,145 @@
+
+namespace espbt = esphome::esp32_ble_tracker;
+
+// "00001000-bed0-0080-aa55-4265644a6574"
+static const espbt::ESPBTUUID BEDJET_SERVICE_UUID = espbt::ESPBTUUID::from_raw(
+    (uint8_t[]){0x74, 0x65, 0x4a, 0x64, 0x65, 0x42, 0x55, 0xaa, 0x80, 0x00, 0xd0, 0xbe, 0x00, 0x10, 0x00, 0x00});
+
+// "00002000-bed0-0080-aa55-4265644a6574"
+static const espbt::ESPBTUUID BEDJET_STATUS_UUID = espbt::ESPBTUUID::from_raw(
+    (uint8_t[]){0x74, 0x65, 0x4a, 0x64, 0x65, 0x42, 0x55, 0xaa, 0x80, 0x00, 0xd0, 0xbe, 0x00, 0x20, 0x00, 0x00});
+
+// "00002004-bed0-0080-aa55-4265644a6574"
+static const espbt::ESPBTUUID BEDJET_COMMAND_UUID = espbt::ESPBTUUID::from_raw(
+    (uint8_t[]){0x74, 0x65, 0x4a, 0x64, 0x65, 0x42, 0x55, 0xaa, 0x80, 0x00, 0xd0, 0xbe, 0x04, 0x20, 0x00, 0x00});
+
+// "00002001-bed0-0080-aa55-4265644a6574"
+static const espbt::ESPBTUUID BEDJET_NAME_UUID = espbt::ESPBTUUID::from_raw(
+    (uint8_t[]){0x74, 0x65, 0x4a, 0x64, 0x65, 0x42, 0x55, 0xaa, 0x80, 0x00, 0xd0, 0xbe, 0x01, 0x20, 0x00, 0x00});
```
When it's merged, we'll be able to use the new overload added here https://github.com/esphome/esphome/pull/2453/files#diff-c581745ebf09e9c23ad69f4bc9c01c49858759277e3c347afec5b3596509f521R313

jhansche commented on esphome/components/bedjet/bedjet.h:
```diff
@@ -0,0 +1,134 @@
+        climate::CLIMATE_MODE_DRY,
+    });
+
+    // TODO: decide if we stick with low/med/high, or use 5% increments
+    traits.set_supported_fan_modes({
+        climate::CLIMATE_FAN_HIGH,
+        climate::CLIMATE_FAN_MEDIUM,
+        climate::CLIMATE_FAN_LOW,
+    });
+    // It would be better if we had a slider for the fan modes.
+    // FIXME: this set is not ordered by insertion, it is sorted alphabetically...
+    traits.set_supported_custom_fan_modes({"5%",  "10%", "15%", "20%", "25%", "30%", "35%", "40%", "45%", "50%",
+                                           "55%", "60%", "65%", "70%", "75%", "80%", "85%", "90%", "95%", "100%"});
```
Would be interested in suggestions here. The device can set fan speed in 5% increments (5-100%), so low/med/high is a bit too simplistic for full control.

But setting 20 fan modes for each step also sounds unrealistic. Plus doing this, it sorts them alphabetically, so we see `"10", "100", "15", "20", ..  "5", "50", ..` etc. And that doesn't make sense when viewing the fan presets in HA.

Ideally, it would be great to have a uint8_t, or 0..100 fan speed, and a fan speed slider, like what we have in https://esphome.io/components/fan/speed.html. So that we can set a min/max/step, and let the slider do the work. But I don't know if it's possible to have a single component support both the Climate component and the Fan component at the same time.

jesserockz replied:
I guess instead of having fan mode for the climate, you could just expose a fan entity beside the climate entity to control the fan setting independently?

jhansche replied:
Wouldn't that then have to maintain its own separate ble notify status packet, in order to receive current fan speed, and control the fan speed? I was trying to avoid having to maintain the incoming ble notifications in multiple places 🤔

jesserockz replied:
You could have the logic in a single base component/hub

jesserockz replied:
A quick thought, have you tried prepending them with 0, not the best solution, but they would be in order at least

jhansche replied:
I guess BleClient will maintain the single upstream ble connection, and just dispatches notifications to each.  But that means if one disabled notifications it would impact the other device as well. I already have to manage notifications separately from the ble client due to a bug.

I could make it configurable where if you want more control you can add a second BedJet/fan component for the same ble client, and disable the climate fan modes; otherwise you just get simple low/med/high.

jhansche replied:
> A quick thought, have you tried prepending them with 0, not the best solution, but they would be in order at least

I haven't yet, but since it's 5-100, we'd have to have 3 digits 🤔  `005`, `010`, .. `095`, `100`. I agree that it would solve the ordering issue, but would not be my first choice.

> You could have the logic in a single base component/hub

Do you have any examples off the top of your head, of a component that does something like this: a base hub idea that manages the communication with the device, and separate device-type components or platforms that share the single hub/connector? Sounds like a good idea to support this.

I do already have a separate `BedjetCodec` class that receives and caches the notification packet, and also prepares the outgoing command packets that will be sent to the device.

jhansche replied:
I got around to testing this again finally, and while the 0-prefix does work, I also found that using space as the prefix works as well, and gives a better appearance in the UI. In fact, the UI appears to trim the leading spaces, so it looks like I expect it to, but maintains the leading spaces when sending the ClimateCall, so I'm able to do the reverse lookup still.

Given that, I'm going to remove the low/med/high, and keep the space-prefixed increments.

jhansche committed:
Fix new clang issues

jhansche committed:
Fix custom fan mode sorting

The UI presents fan modes in alphabetical order, which breaks the
ability to list numeric increments as the fan mode.

By prefixing each increment with leading spaces, the UI continues to
show the numbers as intended, but in the correct order. We still have
to compare the current custom fan mode using leading spaces.

jhansche commented on esphome/components/bedjet/bedjet_const.h:
```diff
@@ -0,0 +1,78 @@
+  CMD_SET_TIME = 0x8,
+};
+
+#define BEDJET_FAN_STEP_NAMES_ \
+  { \
+    "  5%", " 10%", " 15%", " 20%", " 25%", " 30%", " 35%", " 40%", " 45%", " 50%", " 55%", " 60%", " 65%", " 70%", \
+        " 75%", " 80%", " 85%", " 90%", " 95%", "100%" \
+  }
+
+static const char *const BEDJET_FAN_STEP_NAMES[20] = BEDJET_FAN_STEP_NAMES_;
+static const std::string BEDJET_FAN_STEP_NAME_STRINGS[20] = BEDJET_FAN_STEP_NAMES_;
+static const std::set<std::string> BEDJET_FAN_STEP_NAMES_SET BEDJET_FAN_STEP_NAMES_;
```
This is messy. I want to have a single source of truth for the fan mode names (5-100%), but climate_traits needs it to be `std::set<std::string>`, and I also need to be able to infer the "indexof" the incoming string, which is not straight forward in std::set.

It's been about 20 years since my C++ days, so I'm open to suggestions here!

jesserockz replied:
I think for now the space prefix is probably ok :+1:

jhansche committed:
Fix new clang errors

jhansche committed:
Update code comments

* Clarify FIXME after switching to 5% fan speeds
* Change TODO for biodata preset buttons into plain comment

jesserockz commented on esphome/components/bedjet/bedjet.cpp:
```diff
@@ -0,0 +1,656 @@
+      pkt = this->codec_->get_button_request(BTN_M2);
+    } else if (preset == "M3") {
+      pkt = this->codec_->get_button_request(BTN_M3);
+    } else if (BEDJET_DEBUG) {
+      // When BEDJET_DEBUG is enabled, we can optionally offer some additional commands via "presets"
+      if (preset == "Update Firmware") {
+        // For debugging, this can check for a firmware update
+        pkt = this->codec_->get_button_request(MAGIC_UPDATE);
+      } else if (BEDJET_DEBUG && preset == "Debug On") {
+        pkt = this->codec_->get_button_request(MAGIC_DEBUG_ON);
+      } else if (BEDJET_DEBUG && preset == "Debug Off") {
+        pkt = this->codec_->get_button_request(MAGIC_DEBUG_OFF);
```
You should expose an optional `button` for each of these commands. (or a switch for debug on/off)
Having code paths that are locked behind changing a variable in code is not ideal

jhansche replied:
Great idea!

Unfortunately I don't really know how I can expose multiple domain entities from a single controller/hub device. You had suggested that previously, when talking about the fan (wanting to support fan control using a slider instead of low/med/high or 5-100% increments); it would be the same issue here: in order to support that, I would need to have two separate domain components, each using a single shared ble_client connection, and I don't know how to accomplish that. I had asked for recommendations for how we can achieve that. Even if it's just a nudge in the right direction - maybe another component that does that, or a specific esphome API that I can grep for to see how that can be done. Since this component inherits both the Climate component and BLEClientNode, it seems like it would be more difficult to split the BLE client from the Climate component.

It also kind of looks like the component's use of `BedjetCodec` object is not designed for sharing, based on the changes to `unique_ptr()` and `make_unique()`. But I admit it's been a long time since I've touched C++, and I'm not familiar enough with what those apis would do in the context of esphome. I only added those in one of my dev rebase commits, because the same change was added in other components (like `anova` which is what this implementation was based on).

It was my understanding that if I created a new domain component (`fan` or `button` or `switch`), utilizing a shared ble_client_id like I do in the climate component, then each component would be creating its own instance of the `BedjetCodec`, which means each one would be handling the BLE state independently, which sounds like a waste of resources.

At least in this case (compared to the `fan` example), we don't really have a need for maintaining any realtime BLE notification state, because it's just a simple ble packet write if we change it to a button or service. So that could happen without needing a new BedjetCodec instance (or at least it could avoid setting up status notifications).

---

In the near term, I could remove the debug flag entirely, since I'm pretty sure we don't really need this. I found it useful at the time, to be able to restart the BedJet device. But after fixing the ble notify issues, I haven't had a need to do that recently.

jhansche replied:
Given that I'm not sure the exact right approach for splitting the ble_client apart from the `climate` component, and these features aren't really necessary for using the component, we decided to remove the debug flag for now, and I can continue researching how to split that at a later time.

jhansche replied:
I was able to make a button for this, by exposing a public method on the Bedjet class, and adding a new template button that uses an on_press lambda to invoke the method directly. Do you think that's a viable alternative?
```
button:
  - platform: template
    name: "Update Bedjet Firmware"
    on_press:
      then:
        - lambda: |-
            id(my_bedjet_climate_id).upgrade_firmware();
```
```
/** Attempts to check for and apply firmware updates. */
void Bedjet::upgrade_firmware() {
  auto *pkt = this->codec_->get_button_request(MAGIC_UPDATE);
  auto status = this->write_bedjet_packet_(pkt);

  if (status) {
    ESP_LOGW(TAG, "[%s] esp_ble_gattc_write_char failed, status=%d", this->parent_->address_str().c_str(), status);
  }
}
```
Then this template button config could be added to the [docs PR](https://github.com/esphome/esphome-docs/pull/1508/files#diff-49c9c76ae2f8342da70cc8d5b97e9c6f43055b3b2995519ccea698fd510c8bce), alongside other tips like the `ota.on_begin` automation idea.

jhansche committed:
Remove build-time debugging

jhansche committed:
Add public method to attempt firmware upgrade

jesserockz reviewed (changes requested):
I just realised that the `time` is optional in config and therefore you need to wrap every function and variable that has to do with time in 
```c++
#idfef USE_TIME
...
#endif
```

Including the `#includes` 

Otherwise if someone has no `time` entry in YAML at all then it won't compile.

jhansche committed:
Guard use of realtime clock

jhansche commented:
Nice catch re `USE_TIME`. I thought I tested that, but I guess not. Is there any reason my test.yaml changes didn't point this out? Does it just happen to be in a file that also defines a time id, so that's why it worked?

jhansche commented:
Looks like these check failures are unrelated to my change :(
```
Error: esphome/components/modbus_controller/select/__init__.py: Please format this file with the black formatter
```
```
Error: /home/runner/work/esphome/esphome/esphome/components/max9611/max9611.cpp:16:1: error: 'const' type qualifier on return type has no effect [clang-diagnostic-ignored-qualifiers,-warnings-as-errors]
const float gain_to_lsb(MAX9611Multiplexer gain) {
^
Error: /home/runner/work/esphome/esphome/esphome/components/max9611/max9611.cpp:16:1: error: return type 'const float' is 'const'-qualified at the top level, which may reduce code readability without improving const correctness [readability-const-return-type,-warnings-as-errors]
```

jhansche committed:
Merge remote-tracking branch 'origin/dev' into bedjet

* origin/dev:
  Only get free memory size from internal (#3259)
  Remove stray define (#3260)
  Add Mopeka BLE and Mopeka Pro Check BLE Sensor (#2618)

jhansche committed:
Merge remote-tracking branch 'origin/dev' into bedjet

* origin/dev:
  Add optimistic config flag to modbus select. (#3267)
  Bump docker dependencies (#3281)
  Bump click from 8.0.3 to 8.0.4 (#3248)
  Bump zeroconf from 0.38.3 to 0.38.4 (#3257)
  Bump pytest-asyncio from 0.18.1 to 0.18.2 (#3262)
  Bump version to 2022.4.0-dev
  Add visual step/min/max for webserver climate (#3275)
  Bump esphome-dashboard to 20220309.0 (#3277)
  Add entity_category_diagnostics to SGP30 baseline sensors (#3272)
  Webserver v2 (#2688)

Tuckie commented:
Any updates on this?

Tuckie commented:
Thank you @jesserockz!

kquinsland commented:
> Thank you @jesserockz!

And thank you to @jhansche . Perfect timing as summer is coming here....
